src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIResponseChatClient.cs

```diff
@@ -0,0 +1,582 @@
+    {
+        get => _toolCallJsonSerializerOptions;
+        set => _toolCallJsonSerializerOptions = Throw.IfNull(value);
+    }
+
+    /// <inheritdoc />
+    object? IChatClient.GetService(Type serviceType, object? serviceKey)
+    {
+        _ = Throw.IfNull(serviceType);
+
+        return
+            serviceKey is not null ? null :
+            serviceType == typeof(ChatClientMetadata) ? _metadata :
+            serviceType == typeof(OpenAIResponseClient) ? _responseClient :
+            serviceType.IsInstanceOfType(this) ? this :
+            null;
+    }
+
+    /// <inheritdoc />
+    public async Task<ChatResponse> GetResponseAsync(
+        IEnumerable<ChatMessage> messages, ChatOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(messages);
+
+        // Convert the inputs into what OpenAIResponseClient expects.
+        var openAIResponseItems = ToOpenAIResponseItems(messages, ToolCallJsonSerializerOptions);
+        var openAIOptions = ToOpenAIResponseCreationOptions(options);
+
+        // Make the call to the OpenAIResponseClient.
+        var openAIResponse = (await _responseClient.CreateResponseAsync(openAIResponseItems, openAIOptions, cancellationToken).ConfigureAwait(false)).Value;
+
+        // Convert and return the results.
+        ChatResponse response = new()
+        {
+            ResponseId = openAIResponse.Id,
+            CreatedAt = openAIResponse.CreatedAt,
+            FinishReason = ToFinishReason(openAIResponse.IncompleteStatusDetails?.Reason),
+            Messages = [new(ChatRole.Assistant, (string?)null)],
+            ModelId = openAIResponse.Model,
+            Usage = ToUsageDetails(openAIResponse),
+        };
+
+        if (!string.IsNullOrEmpty(openAIResponse.EndUserId))
+        {
+            (response.AdditionalProperties ??= [])[nameof(openAIResponse.EndUserId)] = openAIResponse.EndUserId;
+        }
+
+        if (openAIResponse.Error is not null)
+        {
+            (response.AdditionalProperties ??= [])[nameof(openAIResponse.Error)] = openAIResponse.Error;
+        }
+
+        if (openAIResponse.OutputItems is not null)
+        {
+            ChatMessage message = response.Messages[0];
+            foreach (ResponseItem outputItem in openAIResponse.OutputItems)
+            {
+                switch (outputItem)
+                {
+                    case MessageResponseItem messageItem:
+                        message.RawRepresentation = messageItem;
+                        message.Role = ToChatRole(messageItem.Role);
+                        (message.AdditionalProperties ??= []).Add(nameof(messageItem.Id), messageItem.Id);
+                        message.Contents = ToAIContents(messageItem.Content);
+                        break;
+
+                    case FunctionCallResponseItem functionCall:
+                        response.FinishReason ??= ChatFinishReason.ToolCalls;
```

Comment:
Out of interest, doesn't `ToFinishReason(openAIResponse.IncompleteStatusDetails?.Reason)` already give us the finish reason in the case of tool calls? What makes it necessary to set this finish reason explicitly instead of just keeping whatever finish reason was given (or none)?

Reply:
> doesn't ToFinishReason(openAIResponse.IncompleteStatusDetails?.Reason) already give us the finish reason in the case of tool calls?

No, there's no tool calls reason. It's just null.